src/Storages/MergeTree/MutateTask.cpp

```diff
@@ -2223,18 +2479,111 @@ bool MutateTask::prepare()
+                    bool is_subset_of_groupby = std::all_of(
+                        update_where_columns.begin(),
+                        update_where_columns.end(),
+                        [&](const auto & col) { return stored_columns.contains(col); });
+
+                    if (!is_subset_of_groupby)
+                    {
+                        if ((*ctx->data->getSettings())[MergeTreeSetting::lightweight_mutation_projection_mode]
+                            == LightweightMutationProjectionMode::REBUILD)
+                            ctx->materialized_projections.insert(proj_desc.name);
+
+                        continue;
+                    }
+                }
+                else
+                {
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown projection type");
+                }
+
+                ctx->materialized_projections.erase(proj_desc.name);
+
+                ctx->projection_mutation_contexts.emplace_back(std::make_shared<ProjectionMutationContext>());
+                auto mutation_context = ctx->projection_mutation_contexts.back();
+
+                mutation_context->name = proj_desc.name;
+                mutation_context->metadata_snapshot = proj_desc.metadata;
+                mutation_context->source_part = projections_name_and_part.find(proj_desc.name)->second;
+
+                projection_commands.emplace_back(MutationCommands{});
+                auto & commands= projection_commands.back();
+
+                MutationHelpers::splitAndModifyMutationCommands(
+                    mutation_context->source_part,
+                    proj_desc.metadata,
+                    alter_conversions,
+                    ctx->commands_for_part,
+                    commands,
+                    ctx->for_file_renames,
+                    ctx->log);
+
+                chassert(!commands.empty());
+
+                auto projection_interpreter = std::make_unique<MutationsInterpreter>(
+                    *ctx->data, mutation_context->source_part, alter_conversions,
+                    proj_desc.metadata, commands,
+                    proj_desc.metadata->getColumns().getNamesOfPhysical(), context_for_reading, settings);
+
+                mutation_context->mutating_pipeline_builder = projection_interpreter->execute();
+                mutation_context->updated_header = projection_interpreter->getUpdatedHeader();
+
+                mutation_context->kind = MutationHelpers::haveMutationsOfDynamicColumns(
+                    mutation_context->source_part, ctx->commands_for_part)
+                    || !isWidePart(mutation_context->source_part)
+                    || !isFullPartStorage(mutation_context->source_part->getDataPartStorage())
+                    || (projection_interpreter && projection_interpreter->isAffectingAllColumns())
+                    ? ProjectionMutationContext::MutationKind::ALL : ProjectionMutationContext::MutationKind::SOME;
```

Comment:
By the way, why not just create a list of `MutateTask`s for projections inside a `MutateTask`? You recreated much of the logic of a mutate task while iterating over projection contexts. A similar logic is in the `MergeTask` where the main `MergeTask` has a list with tasks (`tasks_for_projections`) which are regular `MergeTask`s.